expression/builtin_json.go

```diff
@@ -51,471 +59,448 @@ var (
-		return d, errors.Trace(err)
-	}
-	d.SetMysqlJSON(j)
-	return
-}
-
-// JSONMerge is for json_merge builtin function.
-func JSONMerge(args []types.Datum, sc *variable.StatementContext) (d types.Datum, err error) {
-	if argsAnyNull(args) {
-		return d, nil
-	}
-	jsons := make([]json.JSON, 0, len(args))
-	for _, arg := range args {
-		j, err := datum2JSON(arg, sc)
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-		jsons = append(jsons, j)
-	}
-	d.SetMysqlJSON(jsons[0].Merge(jsons[1:]))
-	return
-}
-
-// JSONObject creates a json from an ordered key-value slice. It retrieves 2 arguments at least.
-func JSONObject(args []types.Datum, sc *variable.StatementContext) (d types.Datum, err error) {
-	if len(args)&1 == 1 {
-		err = ErrIncorrectParameterCount.GenByArgs(ast.JSONObject)
-		return
-	}
-	var jsonMap = make(map[string]json.JSON, len(args)>>1)
-	var keyTp = types.NewFieldType(mysql.TypeVarchar)
-	for i := 0; i < len(args); i += 2 {
-		if args[i].Kind() == types.KindNull {
-			err = errors.New("JSON documents may not contain NULL member names")
-			return
-		}
-		key, err := args[i].ConvertTo(sc, keyTp)
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-		value, err := args[i+1].ToMysqlJSON()
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-		jsonMap[key.GetString()] = value
-	}
-	j := json.CreateJSON(jsonMap)
-	d.SetMysqlJSON(j)
-	return
-}
-
-// JSONArray creates a json from a slice.
-func JSONArray(args []types.Datum, sc *variable.StatementContext) (d types.Datum, err error) {
-	jsons, err := createJSONFromDatums(args)
-	if err != nil {
-		return d, errors.Trace(err)
-	}
-	j := json.CreateJSON(jsons)
-	d.SetMysqlJSON(j)
-	return
-}
+	// Type of JSON value.
+	_ builtinFunc = &builtinJSONTypeSig{}
+	// Unquote JSON value.
+	_ builtinFunc = &builtinJSONUnquoteSig{}
+	// Create JSON array.
+	_ builtinFunc = &builtinJSONArraySig{}
+	// Create JSON object.
+	_ builtinFunc = &builtinJSONObjectSig{}
+	// Return data from JSON document.
+	_ builtinFunc = &builtinJSONExtractSig{}
+	// Insert data into JSON document.
+	_ builtinFunc = &builtinJSONSetSig{}
+	// Insert data into JSON document.
+	_ builtinFunc = &builtinJSONInsertSig{}
+	// Replace values in JSON document.
+	_ builtinFunc = &builtinJSONReplaceSig{}
+	// Remove data from JSON document.
+	_ builtinFunc = &builtinJSONRemoveSig{}
+	// Merge JSON documents, preserving duplicate keys.
+	_ builtinFunc = &builtinJSONMergeSig{}
+)
 
 type jsonTypeFunctionClass struct {
 	baseFunctionClass
 }
 
 type builtinJSONTypeSig struct {
-	baseBuiltinFunc
+	baseStringBuiltinFunc
 }
 
 func (c *jsonTypeFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinJSONTypeSig{newBaseBuiltinFunc(args, ctx)}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpJSON)
+	bf.tp.Flen = 51 // max length of UNSIGNED INTEGER.
+	args[0].GetType().Decimal = castJSONDirectly
```

Comment:
why change `Decimal` of return type for `arg[0]` ?

Reply:
Because we should do cast here with `CAST`, not `COERCE`.